### PR TITLE
Add Census resource API and implementation.

### PR DESCRIPTION
**Pull request by a-veitch**
Changes in this PR, in approximate order of significance:
* Add core Resource API, implementation and tests.
* Rename two of the Census proto messages to better reflect existing naming in other google systems.
* Add base resource definitions for latency resources, and add them when census is initialized.
* Remove the unimplemented/placeholder census stats API functions from census.h. Will re-add as they are implemented.
* Extend the census initialization function to return features enabled, with corresponding cleanup in implementation.

### CODE REVIEW

**Comment by a-veitch** [comment]
Remaining test failures look unrelated (macos port server failure).

File: src/core/ext/census/base_resources.c [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+#include <string.h>
+
+#include <grpc/census.h>
+#include <grpc/support/log.h>
+
+#include "base_resources.h"
+#include "gen/census.pb.h"
+#include "third_party/nanopb/pb_encode.h"
+
+// Add base RPC resource definitions for use by RPC runtime.
```
Comment (bogdandrutu) [inline src/core/ext/census/base_resources.c]: Should this be C89 comments?

File: src/core/ext/census/resource.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+// Census-internal resource definition and manipluation functions.
```
Comment (bogdandrutu) [inline src/core/ext/census/resource.h]: C89 comments in the header?

**Comment by bogdandrutu** [comment]
Please fix tests.

**Comment by a-veitch** [comment]
AFAIKT, all the test failures are unrelated - macos still failing on port_server (basic tests) and some socket failure in one of the interop tests. PTAL.

File: include/grpc/census.h [hunk 3]
```diff
@@ -430,109 +432,51 @@ CENSUSAPI int census_get_trace_record(census_trace_record *trace_record);
...
-   * Metric: Each measurement is for a single metric. Examples include RPC
-     latency, CPU seconds consumed, and bytes transmitted.
-   * View: A view is a combination of a metric, a tag set (in which the tag
-     values are regular expressions) and a set of aggregations. When a
-     measurement for a metric matches the view tags, it is recorded (for each
-     unique set of tags) against each aggregation. Each metric can have an
-     arbitrary number of views by which it will be broken down.
+   * Resource: Users record measurements for a single resource. Examples
+     include RPC latency, CPU seconds consumed, and bytes transmitted.
+   * Aggregation: An aggregation of a set of measurements. Census supports the
```
Comment (bogdandrutu) [inline include/grpc/census.h]: Are we not going to support SUM?

File: src/core/ext/census/resource.c [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+// Number of defined resources
+static size_t n_defined_resources = 0;
+
+void initialize_resources(void) {
+  gpr_mu_init(&resource_lock);
+  gpr_mu_lock(&resource_lock);
+  GPR_ASSERT(resources == NULL && n_resources == 0 && n_defined_resources == 0);
+  // 8 seems like a reasonable size for initial number of resources.
+  n_resources = 8;
```
Comment (bogdandrutu) [inline src/core/ext/census/resource.c]: Do you need an initial size for the number of resources? What about leaving this 0?

File: src/core/ext/census/resource.c [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+  resources = gpr_malloc(n_resources * sizeof(resource *));
+  memset(resources, 0, n_resources * sizeof(resource *));
+  gpr_mu_unlock(&resource_lock);
+}
+
+// Delete a resource given it's ID. Must be called with resource_lock held.
```
Comment (bogdandrutu) [inline src/core/ext/census/resource.c]: Update the comment to say that the resource must not be NULL.

File: src/core/ext/census/resource.c [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+  if (rid >= 0 && (size_t)rid < n_resources && resources[rid] != NULL) {
+    delete_resource_locked((size_t)rid);
+  }
+  gpr_mu_unlock(&resource_lock);
+}
+
+int32_t census_resource_id(const char *name) {
+  gpr_mu_lock(&resource_lock);
+  for (int32_t id = 0; (size_t)id < n_resources; id++) {
+    if (resources[id] != NULL) {
```
Comment (bogdandrutu) [inline src/core/ext/census/resource.c]: if (resources[id] != NULL && strcmp(resources[id]->name, name) == 0) {
  ...
}

File: include/grpc/census.h (hunk 3 above)
Comment (a-veitch) [inline include/grpc/census.h]: > Are we not going to support SUM?

Sums are part of the Distribution aggregation.

File: src/core/ext/census/resource.c (hunk 4 above)
Comment (a-veitch) [inline src/core/ext/census/resource.c]: Left it 0. Now only allocates space when first resource is added.

File: src/core/ext/census/resource.c (hunk 5 above)
Comment (a-veitch) [inline src/core/ext/census/resource.c]: Done.

File: src/core/ext/census/resource.c (hunk 6 above)
Comment (a-veitch) [inline src/core/ext/census/resource.c]: Done

**Pull request merged**